### PR TITLE
More package linking in inline code

### LINKED ISSUE #27: More package linking in inline code
Could

```markdown
Check out the `{downlit}` package!
```

become

```markdown
Check out the [downlit](url) package!
```
?

### PR DESCRIPTION
**Pull request by maelle**
Fix #27

### CODE REVIEW

**Comment by maelle** [comment]
This doesn't work for html yet. :-)

**Comment by maelle** [comment]
@hadley how much re-factoring would be ok for making this work for HTML? :thinking: At the moment `tweak_children(x, xpath_inline, autolink, replace = "contents")` understandably only replaces the contents whereas my proposed change changes a node into plain text with a link, from code.

File: R/downlit-md.R [hunk 1]
```diff
@@ -126,7 +126,22 @@ transform_code <- function(x, version) {
     lapply(x, transform_code, version = version)
   } else {
     if (x$t == "Code") {
-      href <- autolink_url(x$c[[2]])
+      # packages à la {pkgname}
+      # regex adapted from https://github.com/r-lib/usethis/blob/d5857737b4780c3c3d8fe6fb44ef70e81796ac8e/R/description.R#L134
+      if(grepl("^\\{[a-zA-Z][a-zA-Z0-9.]+\\}$", x$c[[2]])) {
```
Comment (hadley) [inline R/downlit-md.R:131]: Can you please put this in a separate function?  (i.e. either something like `is_package_name()`  or `extract_curly_package()` that returns `NA` if its not found)

File: R/downlit-md.R [hunk 2]
```diff
@@ -126,7 +126,22 @@ transform_code <- function(x, version) {
...
+        x$c[[2]] <- sub(
```
Comment (hadley) [inline R/downlit-md.R:132]: I think this would be easier to understand if done in two separate steps.

**Comment by maelle** [comment]
Thanks, I'll dig into this again next week. :eyes:

**Comment by maelle** [comment]
* I'll need to update tests once I rebase the PR.

* I'm a bit stuck regarding tweaking the parent in the case where there is no link to be added (in the case where `<code>{downlit}</code>` is replaced with `<a>downlit</a>`, it is easy because one simply replaces a node). I thought it'd be smart to replace the code node with a place-holder node that I'd then strip. I could however not find a solution for stripping a node whilst keeping its content. The current solutions "works" but a) it uses regular expressions to strip the place-holder node tags b) the output of `xml2::xml_contents()` is surprising to me.

**Comment by maelle** [comment]
I'll mark this as ready for review because I'm still stuck about the second point above. I wish one could remove a node by merging its text into the parent node... in an elegant way. :grimacing:

**Comment by hadley** [comment]
Could you please give me a simple reprex of the node stripping problem? (i.e. something just involving xml2). Then I can take a stab at it.

**Comment by maelle** [comment]
No xml2, just input and output, if it is not enough I can add more on Wednesday. How to transform

```xml
<p> blabla <placeholder> pkg </placeholder> other blabla </p>
```

Into

```xml
<p>blabla pkg other blabla</p>
```

That, or maybe the whole way I am trying to add feature is wrong 🙂 (or maybe one could keep a span since it could be a target for styling 🤷‍♀️ )

**Comment by hadley** [comment]
Here's one way to tackle it:

``` r
library(xml2)

xml <- read_xml("<p> blabla <placeholder> pkg <b>bold</b> </placeholder> other blabla </p>")
placeholder <- xml_find_first(xml, ".//placeholder")

contents <- xml_contents(placeholder)
for(node in contents) {
  xml_add_sibling(placeholder, node, .where = "before")
}
xml_remove(placeholder)

as.character(xml, options = "no_declaration")
#> [1] "<p> blabla  pkg <b>bold</b>  other blabla </p>\n"
```

<sup>Created on 2020-09-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

I don't know if it's really any better than your approach, but keeping everything in xml2's data structure makes me feel more confident that we won't accidentally corrupt the output. If you decide to take this approach, make sure to keep this as a separate function, and test edge cases (e.g. empty placeholder, placeholder inside placeholder, ...).

**Comment by maelle** [comment]
The test failures seem to be related to pkgdown?

**Comment by hadley** [comment]
I simplified the logic quite a bit; did I make some mistake?

(Running down the pkgdown bug now)

**Comment by hadley** [comment]
(pkgdown issue is https://github.com/r-lib/pkgdown/pull/1402; will be fixed once pkgdown goes to CRAN)

File: tests/testthat/test-downlit-html.R [hunk 3]
```diff
@@ -7,3 +7,17 @@ test_that("can highlight html file", {
...
 })
+
+test_that("special package string gets linked", {
+  # TODO: convert to snapshot tests
+  html <- xml2::read_xml("<p>before <code>{downlit}</code> after</p>")
+  downlit_html_node(html)
+  expect_equal(length(xml2::xml_find_all(html, ".//a")), 1)
+  expect_equal(length(xml2::xml_find_all(html, ".//code")), 0)
+
+  # But only when it's a real package
```
Comment (maelle) [inline tests/testthat/test-downlit-html.R:18]: so if it's not linkable, it's not transformed at all, correct? The line I had added in the README is therefore a bit wrong.

File: README.md [hunk 4]
```diff
@@ -38,6 +38,7 @@ The following forms of inline code are recognized and automatically linked:
 * `help("fun")`, `help("fun", package = "package")`, `help(package = "package")`.
 * `vignette("name")`, `vignette("name", package = "package")`.
 * `library(package)`, `require(package)`, `requireNamespace("package")`.
+* `{package}` gets linked (if possible) _and formatted as plain text_.
```
Comment (maelle) [inline README.md:41]:
```suggestion
* `{package}` gets linked if possible. If an URL is found, it is linked and formatted as plain text. If no URL is found, it stays as is.
```

File: tests/testthat/test-downlit-html.R (hunk 3 above)
Comment (maelle) [reply tests/testthat/test-downlit-html.R:18]: but it's not consistent with what happens in `downlit_md()` actually?

File: README.md (hunk 4 above)
Comment (maelle) [reply README.md:41]: at the moment different behaviors for Markdown and HTML.

File: tests/testthat/test-downlit-html.R (hunk 3 above)
Comment (hadley) [reply tests/testthat/test-downlit-html.R:18]: Right, why would you reformat it if it's not a link to a known package?

Comment (hadley) [reply tests/testthat/test-downlit-html.R:18]: But this does explain why my code is much simpler than yours 😄

Comment (maelle) [reply tests/testthat/test-downlit-html.R:18]: I have no idea why I did that (it seemed logical at the time?!) and am currently editing the md code :see_no_evil:

**Comment by maelle** [comment]
This is a new-to-me folder

[image](https://lockedata.github.io/cransays/articles/dashboard.html)

**Comment by maelle** [comment]
So, pkgdown is now on CRAN, and I re-launched checks with a commit, but tests still fail. If I follow correctly it's because testthat pkgdown website hasn't been rebuilt since you fixed the pkgdown bug @hadley?

And do you have further comments on this PR?

**Comment by hadley** [comment]
Just started pkgdown re-running on testthat, but I think it's fine to merge this now. Thanks!
